### ISSUE TITLE
`update` command

### ISSUE THREAD

**Issue by SevInf** [gemini-testing/gemini#36]
Should work as follows:

- by default, gather new images only for missing states
- if passed `--force` flag, update all images.

`gather` should be deprecated after this command is implemented.

**Comment by SevInf** [comment]
No need for this command, `GUI` solves all the problems which we intended to solve with it.

**Comment by fenduru** [comment]
Would it be possible to include programatic hooks for this functionality? I'm looking to create a gulp task to interface with gemini, and adding/updating reference shots seems to be a challenge currently.

**Comment by fenduru** [comment]
@SevInf any thoughts on this? I would be glad to work on the enhancement

**Comment by SevInf** [comment]
@fenduru sorry, I'm no longer leading this project, better ask @SwinX or @sipayRT

**Comment by sipayRT** [comment]
@fenduru why don't you use `gather` command for your task? Can you provide more details?

**Comment by fenduru** [comment]
@sipayRT the only problem I have with `gather` is that it re-creates the images. I want to be able to run the tests, see the diff, and if the diff looks correct then accept the new image. But currently I have to run the tests, see the diff, then re-generate the images with `gather`.

In a perfect world, all of these tests would be deterministic, so re-generating the image wouldn't concern me. However, we occasionally run into issues where an element accidentally has hover styling applied, etc, and I wouldn't want to accidentally commit incorrect images to our git repo.

The gemini-gui project looks like it has some of this functionality, however my typical workflow will be to run the tests through gulp, and then look at/accept changed images after that. Running a long-lived server does not fit nicely into my build system.

It seems that currently Gemini doesn't actually save any of the test result images to disk, but if it did, I think it would be nice to have an additional command called `accept` that would replace reference images that have changed. For instance:

```
$ gemini gather
$ git commit -am "save reference images"
$ gemini test
test passes
$ vim my-component.css
$ gemini test
test fails, inspect the diff on the HTML reporter
$ gemini accept
$ git commit -am "update changed images"
```

**Comment by sipayRT** [comment]
what should we do when part of fallen tests would be incorrect? For example, because of grid troubles.

**Comment by fenduru** [comment]
I would expect the person running the tests to verify that all of the images are correct by viewing the HTML report before running the accept command. Personally I would never accept new images if the test run didn't complete without problems, but perhaps it would be useful to support the `--grep` option here too.

**Comment by sipayRT** [comment]
@j0tunn can you show your script here?

**Comment by j0tunn** [comment]
```bash
#!/bin/bash

find gemini-report/images/ | grep ~diff.png | while read -r f
do
    dir=`dirname "$f" | sed 's;gemini-report/images/;;g'`
    browser=`basename "$f" | sed 's;~diff.png;;g'`
    cp "gemini-report/images${dir}/${browser}~current.png" "gemini/screens${dir}/${browser}.png"
done
```

**Comment by fenduru** [comment]
I haven't tried running this script yet, but it looks like the behavior I'm looking for. The problem I see with having a script like that though is that it is coupled with the reporter, which could very reasonably change the way it is structured (since its job is reporting).

How would you feel about having a `gemini/screens/reference` folder containing the accepted images, and a `gemini/screens/pending` folder containing any failed images? This would make implementing an `accept` command (or something similar).

**Comment by fenduru** [comment]
Thinking about this a bit more, and I think my complaint is really that `gather` is overloaded for both new tests, as well as changed tests.

I think that if an `accept` command is created, that `gather` should **only** gather test with no existing references, and the only way to change existing references is by running `test` followed by `accept` (or, of course, by deleting the reference images).

**Comment by j0tunn** [comment]
The main thing is that you **should look** at new screenshots before accepting them, so it is all about reporting images. That is not the gemini's job to store images somewhere between launches. It just launches tests and gives you result. Thats it.
For the purpose of convenient updating images we've created `gemini-gui`.
If for some reason you can't use `gemini-gui` you can use the script I've posted earlier. It's a simple 6-line script, I don't think its a big deal to adopt it in case of some html-reporter (unlikely) breaking change.
And off course there is always another option for you: plugin. I can think of custom reporter that stores the images somewhere and creates the update script based on paths from gemini config. Launch that script after `gemini test` and boom! The job is done.

**Comment by fenduru** [comment]
In the short term I can use that script, but I still don't think this is correct. Based off of what you are saying, then `gemini gather` should be deprecated, since it is not gemini's job to store the results between launches, and this logic should instead be handled by the reporter + script.

The difference between gemini and more typical test runners is that you NEED a reference image in order for the system to work. Creating and updating these reference images is integral to the system, so it seems strange to me that you'd rather this be handled outside of the system.

**Comment by fenduru** [comment]
`gather` isn't quite what I need even when I want to add new tests. The problem is that it not only generates images for new tests, but it also replaces existing reference images for failing tests (without letting you see the changes).

I could work around this, but there is no way to generate new reference images without using `gather` because of [this line](https://github.com/gemini-testing/gemini/blob/f461f6521499a0f77155b01ffefedfb6a32b8208/lib/tester.js#L47).

So when I add a new test and need to generate an image, I have to just hope that nothing else changed when I run `gather`

**Comment by fenduru** [comment]
### Summary of my issue

* `gather` will generate new images for both existing tests AND new tests without a way of choosing between them
* `gather` results do not appear in the HTML reporter, so I have no way of verifying changes to existing images (and no easy way to verify that brand new images are correct)
* `gemini-gui` requires re-running all of the tests rather than just accepting the failures from the previous test run
  - My workflow is to make a code change, then run my test suites (image, unit, etc. tests), repeat.

### Proposed changes

* `test` runs all tests, even if no reference image exists. 
  - Naturally new tests will fail, but I feel this is correct behavior - I've added a test and its result does not match the reference (since the reference doesn't exist).
* HTML reporter shows new images

### What I will do after those changes are made

* Write a reporter that handles saving the previous test-run's images, and gives me the ability to write an `update` command on my own, outside of Gemini.
* I will never run `gather`, since I don't want to update the `screens` folder until I've reviewed the changes

**Comment by j0tunn** [comment]
> test runs all tests, even if no reference image exists.

Looks like current behaviour. No?

> HTML reporter shows new images

Why not? LGTM

**Comment by fenduru** [comment]
So I dug into the code a bit more, and it turns out that my proposed changes already exist - I just needed to set `referenceImageAbsence` to `warning` **doh**

I still very much think that without a complete story for creating **and** updating reference images after verification, that Gemini is lacking. That said, I should be able to add the features on my own via a plugin or reporter.

**Comment by j0tunn** [comment]
From our experience `referenceImageAbsence: warning` is a bad practice. With this setting tests without reference will always pass, and you can miss them. We are thinking on deprecating this feature.

**Comment by fenduru** [comment]
@j0tunn I agree that an error would be better, however currently the only way I can see to get access to the generated image from a reporter is via the `warning` event.

How would you feel about tweaking [this code](https://github.com/gemini-testing/gemini/blob/52ab59a04620cb8280ad5382f072b1bcae5cefcf/lib/runner/state-runner.js#L48) to always fire an event - maybe a `noReferenceImage` event? Or perhaps have the `endTest` event always contain the full information, regardless of success/failure, reference images, etc.

**Comment by SwinX** [comment]
@fenduru `endTest` now will fire only after test complete with result. If you do not have reference, it will not fire. 
In general we may think about passing some additional info to the events.

**Comment by fenduru** [comment]
Right now I have a local modifciation that additionally passes the `referencePath` to the `warning` event. Obviously this is not desirable.

What I need is an event that fires for every failing test (with or without a reference images) that contains the `currentPath`, and the `referencePath` (which might not exist on disk, but I need to know where I should eventually copy the generated image to).

I'd like to file a PR for this, but would like your input on which event (existing event? new event?) you would want this on.
